### ISSUE TITLE
Support server speaks first protocol linkerd2

### ISSUE THREAD

**Issue by barkardk** [linkerd/linkerd2#5384]
## Feature Request
Since 2.9 mTLS for TCP connections is now supported and can be verified using linkerd edges command. 
Server speaks first protocols such as mysql and smtp , still do not work  though.
In linkerd2 slack space there are several mentions that mTLS for TCP connections where server initiates the connection as with mysql and smpt, will be released in version 2.10. 
No issue has ben created around that specifically for interested adopters to follow so as per suggestion from @olix0r  I am now creating a placeholder to mention updates and PRs around that

**Comment by olix0r** [comment]
https://github.com/linkerd/linkerd2/pull/5294 adds support for an `opaque-ports` annotation so that the proxy can discover whether the target workload should be handled opaquely (i.e. for server-first protocols). The proxy-side support for this annotation has already been implemented.

Following #5294, we plan to add a "connection header" to such streams so that we can transport these connections over mTLS. The first draft of that is implemented in https://github.com/linkerd/linkerd2-proxy/pull/740. I'll update this issue as development continues.

**Comment by olix0r** [comment]
proxy API changes: https://github.com/linkerd/linkerd2-proxy-api/pull/53
proxy support for the new API: https://github.com/linkerd/linkerd2-proxy/pull/785
destination controller to provide new API fields: https://github.com/linkerd/linkerd2/pull/5421
proxy release/upgrade: https://github.com/linkerd/linkerd2/pull/5432

**Comment by olix0r** [comment]
The above changes have merged and should be included in the next edge release. This feature still isn't documented, really; so I'll leave this issue open until we've updated our docs on the website.

To quickly summarize for the sake of testing:

You can now specify a `config.linkerd.io/opaque-ports` annotation on workloads. The value of this annotation is a list of container ports (by number/range and/or name). This configures any proxy running on this workload to skip protocol detection on inbound connections to these ports. It is also discovered by other pods when they attempt to send traffic to this port. If there are proxies on both sides of the connection (running the most recent proxy version), then the outbound proxy will transport the connection securely by sending the traffic directly to the target pod's inbound proxy port (4143) with a special protocol header that informs the inbound proxy of the proper target port. This effectively makes server-first protocols client-first for the purposes of proxy-to-proxy transport, which enables us to perform typical mTLS detection.

**Comment by olix0r** [comment]
edge-21.1.1 includes support for server-first protocols via the `opaque-ports` annotation

**Comment by barkardk** [comment]
I have tested this on a myqsql client connected to a mysql galera cluster and I get connection errors 
SO I did the following 
Installed a galera cluster using bitnami helm chart. just the latest with no data in it on a 3 node cluster 
I set these annotations at first in the statefulset hoping that for example the gcomm port would not be an issue
```
config.linkerd.io/opaque-ports: 3306,4567,4568,4444
```
But the first restarted meshed galera pod  started to throw these errors 
```

demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:33 0 [Note] WSREP: (db6537fc-a1ec, 'tcp://0.0.0.0:4567') connection to peer 00000000-0000 with addr tcp://10.96.2.14:4567 timed out, no messages seen in PT3S, socket stats: rtt: 44 rttvar: 22 rto: 200000 lost: 0 last_data_recv: 3000 cwnd: 10 last_queued_since: 3000147457 last_delivered_since: 3000147457 send_queue_length: 0 send_queue_bytes: 0
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:37 0 [Note] WSREP: (db6537fc-a1ec, 'tcp://0.0.0.0:4567') connection to peer 00000000-0000 with addr tcp://10.96.2.14:4567 timed out, no messages seen in PT3S, socket stats: rtt: 50 rttvar: 25 rto: 200000 lost: 0 last_data_recv: 3000 cwnd: 10 last_queued_since: 3000132125 last_delivered_since: 3000132125 send_queue_length: 0 send_queue_bytes: 0
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [Note] WSREP: PC protocol downgrade 1 -> 0
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [Note] WSREP: view((empty))
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] WSREP: failed to open gcomm backend connection: 110: failed to reach primary view: 110 (Connection timed out)
demo-maria-mariadb-galera-2 mariadb-galera 	 at gcomm/src/pc.cpp:connect():160
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] WSREP: gcs/src/gcs_core.cpp:gcs_core_open():220: Failed to open backend connection: -110 (Connection timed out)
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] WSREP: gcs/src/gcs.cpp:gcs_open():1632: Failed to open channel 'galera' at 'gcomm://demo-maria-mariadb-galera-headless.default.svc.cluster.local': -110 (Connection timed out)
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] WSREP: gcs connect failed: Connection timed out
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] WSREP: wsrep::connect(gcomm://demo-maria-mariadb-galera-headless.default.svc.cluster.local) failed: 7
demo-maria-mariadb-galera-2 mariadb-galera 2021-01-19 12:50:39 0 [ERROR] Aborting
```
So I changed the annotation to 
```
        config.linkerd.io/opaque-ports: "3306"
        config.linkerd.io/skip-inbound-ports: 4567,4568,4444
        config.linkerd.io/skip-outbound-ports: 4567,4568,4444
```

And checked that 3306 is not by default skipped in iptables , this is what iptables look like now with 3306 set tpo opaque and the galera ports are skipped 
````
-A PREROUTING -m comment --comment "proxy-init/install-proxy-init-prerouting/1611060851" -j PROXY_INIT_REDIRECT
-A OUTPUT -m comment --comment "proxy-init/install-proxy-init-output/1611060851" -j PROXY_INIT_OUTPUT
-A PROXY_INIT_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 2102 -m comment --comment "proxy-init/redirect-non-loopback-local-traffic/1611060851" -j PROXY_INIT_REDIRECT
-A PROXY_INIT_OUTPUT -m owner --uid-owner 2102 -m comment --comment "proxy-init/ignore-proxy-user-id/1611060851" -j RETURN
-A PROXY_INIT_OUTPUT -o lo -m comment --comment "proxy-init/ignore-loopback/1611060851" -j RETURN
-A PROXY_INIT_OUTPUT -p tcp -m multiport --dports 4567,4568,4444 -m comment --comment "proxy-init/ignore-port-4567,4568,4444/1611060851" -j RETURN
-A PROXY_INIT_OUTPUT -p tcp -m comment --comment "proxy-init/redirect-all-outgoing-to-proxy-port/1611060851" -j REDIRECT --to-ports 4140
-A PROXY_INIT_REDIRECT -p tcp -m multiport --dports 4190,4191,4567,4568,4444 -m comment --comment "proxy-init/ignore-port-4190,4191,4567,4568,4444/1611060851" -j RETURN
-A PROXY_INIT_REDIRECT -p tcp -m comment --comment "proxy-init/redirect-all-incoming-to-proxy-port/1611060851" -j REDIRECT --to-ports 4143
```

Now I installed a separate mysql instance to use as a client and meshes that with the same opaque port setting for port 3306 
also I have to set skip-inbound and oubound port annotations to empty or the default setting will skip meshing 3306 entirely 
so my mysql client looks like this 
```
        config.linkerd.io/opaque-ports: "3306"
        config.linkerd.io/skip-inbound-ports: ""
        config.linkerd.io/skip-outbound-ports: ""
```
If I now try to connect to the galera cluster from the client (I verified first that I had no issue connecting when 3306 was set to be skipped on bith the client and the cluster)  
I get this familiar error
````
kubectl exec -it mysql-5b6f4f5b5-sbcd4 -- mysql -h demo-maria-mariadb-galera -uroot -p58RX7iFuDv -e 'SHOW databases;' -c mysql
Defaulting container name to mysql.
Use 'kubectl describe pod/mysql-5b6f4f5b5-sbcd4 -n default' to see all of the containers in this pod.
mysql: [Warning] Using a password on the command line interface can be insecure.
ERROR 2013 (HY000): Lost connection to MySQL server at 'reading initial communication packet', system error: 0
command terminated with exit code 1
```
I am using 
level=info msg="running version edge-20.11.1"

**Comment by barkardk** [comment]
As I hit send I noticed that due to a misconfig in my local setup , even if I had installed the proper edge release, that is not what I used for testing. 
I would delete my comment but cant seem to so my hasty remark is sadly stuck. 
SO I have now verified the running version  with the above exact same settings  
 level=info msg="running version edge-21.1.2"  
It works but does not look mTLSed. 
I will create a separate issue for this instead of clogging the comment section on a closed issue
